src/server/io_utils.cc

```diff
@@ -0,0 +1,57 @@
+// Copyright 2022, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "server/io_utils.h"
+
+#include "base/flags.h"
+#include "base/logging.h"
+#include "server/error.h"
+
+namespace dfly {
+
+void BufferedStreamerBase::ReportWritten() {
+  buffered_++;
+  // Check and possibly wait for consumer to keep up.
+  waker_.await([this]() { return !IsStalled() || IsStopped(); });
+  // Notify consumer we have new data.
+  waker_.notify();
```

Comment:
should line 18 and line 16 switch order?
you want to notify that new data was writen and than wait for consumer to keep up if needed

Otherwise you wait till data is consumed by consumer in line 16 if IsStalled, and than you notify that new data was added but it already consumed it while producer waited

Reply:
Yes, this is true - if I stalled from a single write then this wont work. It will however in all other cases.

Otherwise, the consumer can't be both stalled and sleeping, because it would have missed the previous notifications then

Reply:
it's very local, and we can make it more clear by adding more comments.

Reply:
I did not understand the answer. why not swap the order?

Reply:
It needs to be swapped, fixed it